Teleop runstop

erelson opened the pull request:
Tested and works. Probably needs some more cleanup.  Based on testing and hand-walk-throughs, the logic appears to be solid, though is quite dense.  Could maybe be clearer if I refactored this into a state machine?

~~Still need to add differentiation between robot types, and probably allow controlling which breakers are monitored in bringup.~~  Might need to shuffle around the launch file stuff. Testing just now, it seems that both fetchs and freights have fetch_bringup installed, but only freights have freight_bringup, so it seems both robots will be able to use the script in its current location.  (reminder: for now this really just targets fetch; freights' software runstop potentially/likely wants the base motors to brake, instead of coasting as disabling the breaker does.)

I made the script disabled by default.  User can then edit /opt/ros/indigo/share/fetch_bringup/launch/include/runstop.launch.xml

erelson commented:
Script now accepts flags so which breakers it affects can be controlled.  Tested and this works from commandline.

Ready for initial review, methinks.

erelson commented:
I've now got dashboard + controller interactions with the software runstop node troubleshot.  Ready for review.

The one thing to be done (other than squashing) is figure out how we want to do the launch files, so as to ease install on robots in the field (those using calibrated.launch files instead of the default).

This same issue applies to gettin the diagnostics_aggregator node running as well.

erelson commented:
!label review

erelson commented:
!needs-review

erelson commented:
Just a note that I've now tested running the launch files via launching fetch/freight/launch (whereas previously I had just presumed that running the script was sufficient testing).  Did discover
- commented out double hyphens (--) don't work in xml
- Now I know why people are using the parse_known_args() method instead of just parse_args...

erelson commented:
Squashed commits.  Anything else we want to do here? I think Michigan is still interested in getting this functionality. (Looks like I typed this up a week or so ago)

erelson commented:
@dbking77 Melonee suggested you review this at some point.

erelson committed:
Software runstop support via controller and topics

Allows breakers to be disabled via software. Re-enabling
can be done via software, or by toggling the physical runstop.

Teleop control and which breakers are controlled are toggled by
flags. Listens for boolean msgs on topic /enable_software_runstop.
(e.g. for interfacing with rqt_fetch_dash)

Script can be run separately, but is integrated with the
robot drivers, and does not by default enable runstopping with
the controller.

Includes unit tests for statemachine-like behavior of software
runstop.

erelson commented:
@mikeferguson I have rebased this on the latest master.

mikeferguson commented:
@erelson -- as discussed -- please remove launch file changes -- also I don't see any CMake changes that actually run the tests or install the script

mikeferguson commented:
@erelson -- actually I misspoke on that -- let me chat with you in the morning about how we should implement that launch file (although the CMake issues are definitely there)

erelson commented:
Updated.  As discussed:
- Not a python package, so getting catkin to run the test without an import error was not doable; added note about where to run nosetests
- Got rid of freight portion
- Made the "-a -b -g -t" flags an arg that is set in fetch.launch
- Fixed a bug I encountered when testing on a freight; check that robot has arm/base during setup to avoid messing with breakers not used on a given robot

Still need to update the docs pull req.

mikeferguson commented:
LGTM
